### DISCUSSION TITLE
Shortcut settings menu

### DISCUSSION THREAD

**Discussion by Tucchhaa** [solvcon/modmesh#412] [Ideas]
Currently modmesh doesn't have a menu to change/set shortcuts.

**Why do we need that?**
Better UX. It will be possible to attach a hotkey to all the functions of the modmesh, which will add convenience and customization. Also, this menu will allow you to see the full functionality of the application.

A good example is camera movement. Even though we already have intuitive hotkeys to move camera, it's not always obvious that to move camera forward you need to press Ctrl+W or Ctrl+Arrow up. Also there are no shortcuts to rotate camera, only by mouse.

Another advantage is architectural. All of the inputs will be concentrated in one place and not spread across the whole code. It will help to avoid conflict shortcuts.

**How it will look like?**
Like in any IDE. 
_Example from CLion:_
<img width="815" alt="image" src="https://github.com/user-attachments/assets/8b093878-53b6-49bd-8884-346e7d6ba775">

_VS code:_
<img width="1465" alt="image" src="https://github.com/user-attachments/assets/fc73eae0-5727-43a2-97b2-d247538dc498">

**What need to discuss?**
1. What keybindings do we have right now?
2. How exactly this menu will look like? Like CLion's or like VS code's or our own design?
3. What milestones can be? The feature is big and it will require a lot of changes in the source code (at least in camera controller, I can't tell about other parts), so we should try to split it into smaller parts

**Comment by Tucchhaa** [comment]
**Keybindings for camera controller**:

When FPS camera active:
WASD - move camera in X and Y
Ctrl+WS OR Ctrl+Arrow up/Arrow down OR Scroll - move camera in Z
LMB + move mouse - rotate camera
LMB + Shift + move mouse - rotate camera slowly

When Orbit camera active:
WASD OR LMB + move mouse - move camera in X and Y
Ctrl+WS OR Ctrl+Arrow up/Arrow down OR Scroll - move camera in Z
LMB + RMB OR Shift - allow movement only in Z axis
RMB + move mouse OR Alt + WASD OR Alt+Arrows - rotate camera around view center

**Reply by tigercosmos** [reply]
It doesn't make sense to use "Scroll" to zoom. Usually, it's "Ctrl+Scroll" in other softwares.

**Reply by tigercosmos** [reply]
Shortcut for command panel:
1. Ctrl + `: focus on the command panel.
2. Ctrl + C or Ctrl + D: terminal the program.

**Reply by Tucchhaa** [reply]
> It doesn't make sense to use "Scroll" to zoom. Usually, it's "Ctrl+Scroll" in other softwares.

Just "scroll" is how Qt works. When camera controller was developed I intended to add only WASD to controls and not change anything else. In the next PR, I will make it "Ctrl+Scroll" 🙂

**Reply by yungyuc** [reply]
Simply scrolling makes more sense to me.  But the most important (and the most difficult) thing is to make behaviors like this configurable.

We should also consider gestures.  For zooming, pinch in/out should also work.

For something really far-out, check out [SpaceMouse](https://3dconnexion.com/us/product/spacemouse-pro-wireless/) or [MetaQuest](https://www.meta.com/tw/en/quest/).  I don't really think we can "plan" for such fancy I/O devices any time soon, but we should dream big.

**Comment by yungyuc** [comment]
A related question for @Tucchhaa: What else do you want to see in the 2D/3D visualizer, but not limited to it?

The keyboard setting manager is a long-term efforts and will need a couple of design iterations.  I'd like to keep it moving along with other interesting work.

**Reply by Tucchhaa** [reply]
I have several of ideas:

* Render coordinate grid
* Change position of view center using UI (not console)
* CI for camera testing with the renderer

However I'm not sure if there's a real need for these

**Reply by Tucchhaa** [reply]
But I think it's better for me to work on other parts of modmesh. I mean not only the visualizer.

 I'm going to check issues and try to solve some) And will work on settings manager along with that

**Reply by yungyuc** [reply]
> I have several of ideas:
> 
> * Render coordinate grid
> * Change position of view center using UI (not console)
> * CI for camera testing with the renderer
> 
> However I'm not sure if there's a real need for these

For-GUI CI is definitely needed.  It calls for in-depth understandings to Qt and the OS graphical stack and will be enriching.

Once we add more mesh and drawing code, the visualizer will be more interesting.  But right now we don't have the code.  Seeking justification makes sense.

> But I think it's better for me to work on other parts of modmesh. I mean not only the visualizer.

It sounds like a good idea.

> I'm going to check issues and try to solve some) And will work on settings manager along with that

Please take a look.  I think having F2F discussions will be good.  I will be in the scisprint next Saturday (7th September).  Will you be there?

**Reply by Tucchhaa** [reply]
> Please take a look. I think having F2F discussions will be good. I will be in the scisprint next Saturday (7th September). Will you be there?

Yes, I'll be there
